### DISCUSSION TITLE
Unit test examples?

### DISCUSSION THREAD

**Discussion by calumsult** [bikeshaving/crank#161] [General]
Just wondering if anyone has done examples of how to unit test the different component types (particularly async components)? Ideally "everything is a function" should make life much simpler. Sorry if this is a question with an obvious answer.

**Comment by brainkim** [comment]
Definitely no obvious answers. There is [crank-testing-library](https://github.com/ryhinchey/crank-testing-library/blob/master/src/index.js) but I haven’t really tried it yet, and it might be out of date. Typically, when I did test frontend components in React (which was seldomly haha) I usually just rendered the component directly and used Jest’s JSDom to make assertions about the result. I’m not actually 100% sure how people do DOM testing, and one thing that’s kinda painful is that async generator components, or stateful components in general are somewhat difficult to test. Now that there are `schedule` and `cleanup` callback methods, maybe that could be used. Alternatively, I have in the back of my mind the idea of creating mock contexts which can be called to imperatively update and unmount components. Lots of design space to explore!

If you have ideas lemme know. And thanks for checking out Crank!
